Add target_path flag option for edr files and logs

noaKurman opened the pull request.

RoiTabach commented:
Looks good, added one small comment 
Do we want to add user-facing documentation for it?

Maayan-s commented:
> Looks good, added one small comment
> Do we want to add user-facing documentation for it?

@noaKurman 
If you don't document no one will use it, so it's not a question 🙃

noaKurman commented:
> > Looks good, added one small comment
> > Do we want to add user-facing documentation for it?
> 
> @noaKurman If you don't document no one will use it, so it's not a question 🙃 Let me know if you need help with docs. You can do it in different PR
@Maayan-s @RoiTabach 
I added documentation in docs/understand-elementary/cli-commands.mdx, is that enough? If not I'll add it in a different PR :)

Maayan-s commented on docs/understand-elementary/cli-commands.mdx:
````diff
@@ -61,6 +61,13 @@ Here is a list of all the available options:
+```
+
+- **Configuration target directory path**
+  The path where all output files will be saved, default will create "edr_files" directory in your path
````
@noaKurman 
Can they change the default? 
If not I would change to:
The path where `edr` should create an `edr_files` directory where all output files will be saved.

noaKurman replied:
Not completely sure what you meant:
You can either run `--target-path ./noa-target` and it would create `noa-target` in the current directory and save all files there, or you can ignore the flag and it would create `edr_files` in your current directory and save all files there, so should I change it to the phrase you suggested? 
Also should I mention it's an absolute path in the doc?
@Maayan-s

Maayan-s replied:
Ok I get it now! 

So I would write like this:
The absolute path where all output files will be saved. If not configured, the default is `./edr_files`.

RoiTabach commented on elementary/config/config.py:
```diff
@@ -69,10 +72,12 @@ def __init__(
         config = self._load_configuration()
 
-        self.target_dir = self._first_not_none(
+        self.target_dir = self.target_dir = self._first_not_none(
```
Typo without side effects but still - let's fix: 

```suggestion
        self.target_dir = self._first_not_none(
```

RoiTabach commented on elementary/utils/log.py:
```diff
@@ -25,7 +28,30 @@ def format(self, record):
+        except Exception:
+            pass
+        finally:
+            os.makedirs(os.path.abspath(os.path.abspath(file_path)), exist_ok=True)
```
we don't need to double `abspath` , right? 
```suggestion
            os.makedirs(os.path.abspath(file_path), exist_ok=True)
```

RoiTabach commented on elementary/utils/log.py:
```diff
@@ -25,7 +28,30 @@ def format(self, record):
+LOG_FILE = os.getcwd() + "/edr_files/edr.log"
+
+
+class CustomerFileHandler(logging.FileHandler):
```
Is this `customer` and not `custom` intentionaly? I don't know English well enough to be sure

haritamar commented on elementary/utils/log.py:
```diff
@@ -36,7 +62,7 @@ def get_console_handler():
 
 def get_file_handler():
-    file_handler = logging.FileHandler(LOG_FILE)
+    file_handler = CustomFileHandler(LOG_FILE, delay=True)
```
I think we should add the file handler dynamically with the correct log file output, initializing it with `LOG_FILE` and then changing it in `emit` feels a bit hacky...

I think it's better that we'll add a line of the format:
```logger.addHandler(get_file_handler(log_path))```

to `elementary/monitor/cli.py` (where `log_path` is constructed from the target path parameter.

noaKurman commented on elementary/utils/log.py:
```diff
@@ -25,7 +28,30 @@ def format(self, record):
+            target_path_flag = "--target-path"
+            ctx = click.get_current_context()
+            ctx_args = ctx.args
+            if target_path_flag in ctx_args:
```
Because in the context of the cli/cli.py it's an arg, but in the context of some other files it's in the params. I guess its based on if it's before or after the `get_cli_properties` function @haritamar

elongl commented on elementary/monitor/cli.py:
```diff
@@ -154,10 +158,12 @@ def get_cli_properties() -> dict:
     full_refresh_dbt_package = params.get("full_refresh_dbt_package")
 
     return {
+        "target_path": target_path,
```
Is this necessary?
What does it affect?

noaKurman replied:
Are you asking if we need it in the return or if we need the `"target_path": target_path` instead of just returning target_path itself?

For the first question - I'm pretty sure that's how we set the params (instead of it being in the args)

noaKurman commented on elementary/config/config.py:
```diff
@@ -31,13 +31,16 @@ class Config:
 
     DEFAULT_CONFIG_DIR = str(Path.home() / ".edr")
 
+    DEFAULT_FILES_PATH = os.getcwd() + "/edr_files"
```
Or asked for "edr_target" since it's compliant with dbt target dir

noaKurman commented on elementary/utils/log.py:
```diff
@@ -25,7 +28,30 @@ def format(self, record):
+LOG_FILE = os.getcwd() + "/edr_files/edr.log"
+
+
+class CustomerFileHandler(logging.FileHandler):
```
Should have been Custom 😅

haritamar commented on elementary/utils/log.py:
```diff
@@ -46,7 +61,10 @@ def get_logger(logger_name):
     return logger
+
+
+def set_root_logger_file_handler(logger_name):
```
I think you should add both handlers only to the root logger, otherwise you'll get duplicate console outputs. 
So maybe we can have `set_root_logger_handlers`?

noaKurman replied:
I actually don't see it log twice to the console so it's interesting
But anyway I'll move the console handler to the root as well, I agree it's more sorted that way

haritamar commented on elementary/cli/cli.py:
```diff
@@ -45,6 +45,7 @@ def format_help(self, ctx, formatter):
         self.format_epilog(ctx, formatter)
 
     def invoke(self, ctx: click.Context) -> Any:
+        set_root_logger_file_handler("elementary")
```
I think it may be good that `set_root_logger_file_handler` will explicitly get the target path rather than relying on the click context.
WDYT about moving the call to `config.py` below and then doing `set_root_logger_file_handler("elementary", target_path)`?

I think it makes sense to do it right after you're creating the directory there.

noaKurman replied:
Hmm I see at that point in the code the config still points to the default edr_files path, maybe you can show me why 
but I can also get it from the click context in the cli.py instead of the log (unless what you dislike here is the context itself and not the location of it)

haritamar commented on elementary/utils/log.py:
```diff
@@ -35,8 +37,21 @@ def get_console_handler():
     return console_handler
 
 
+def get_log_path():
```
See my note above regarding getting the target_path as a parameter, I think it can simplify this function

haritamar commented on elementary/utils/log.py:
```diff
@@ -35,8 +37,21 @@ def get_console_handler():
+        file_path = ctx_args[ctx_args.index(target_path_flag) + 1]
+    finally:
+        os.makedirs(os.path.abspath(file_path), exist_ok=True)
+        return file_path + "/edr.log"
```
it's a good practice in python usually to use `os.path.join(file_path, "edr.log")` rather than adding strings 
(handles well edge cases, like if the path ends or doesn't end with "/")

noaKurman replied:
Ha good to know, I saw both in the code so just used one of them 👍

haritamar commented on .github/workflows/test-github-action.yml:
```diff
@@ -100,4 +100,4 @@ jobs:
         with:
           name: edr.log
-          path: edr.log
+          path: edr_files/edr.log
```
Should this be changed to `edr_target`? (also for other cases below)
